src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -1484,46 +1495,151 @@ public override IOperation VisitUnaryOperator(IUnaryOperation operation, int? ca
-                    Debug.Assert(binOp.LeftOperand.Type.SpecialType == SpecialType.System_Boolean);
-                    Debug.Assert(binOp.RightOperand.Type.SpecialType == SpecialType.System_Boolean);
-
                     // Rewrite (a || b) as ~(~a && ~b)
-                    andOrSense = !andOrSense;
-                    // Fall through
-                    goto case BinaryOperatorKind.ConditionalAnd;
+                    return !sense;
 
                 case BinaryOperatorKind.ConditionalAnd:
-                    Debug.Assert(binOp.LeftOperand.Type.SpecialType == SpecialType.System_Boolean);
-                    Debug.Assert(binOp.RightOperand.Type.SpecialType == SpecialType.System_Boolean);
-
-                    // ~(a && b) is equivalent to (~a || ~b)
-                    if (!andOrSense)
-                    {
-                        // generate (~a || ~b)
-                        return VisitShortCircuitingOperator(binOp, sense: sense, stopSense: sense, stopValue: true, captureIdForResult, fallToTrueOpt, fallToFalseOpt);
-                    }
-                    else
-                    {
-                        // generate (a && b)
-                        return VisitShortCircuitingOperator(binOp, sense: sense, stopSense: !sense, stopValue: false, captureIdForResult, fallToTrueOpt, fallToFalseOpt);
-                    }
+                    return sense;
 
                 default:
                     throw ExceptionUtilities.UnexpectedValue(binOp.OperatorKind);
             }
         }
 
+        private IOperation VisitBinaryConditionalOperator(IBinaryOperation binOp, bool sense, int? captureIdForResult,
+                                                          BasicBlock fallToTrueOpt, BasicBlock fallToFalseOpt)
+        {
+            // ~(a && b) is equivalent to (~a || ~b)
+            if (!CalculateAndOrSense(binOp, sense))
+            {
+                // generate (~a || ~b)
+                return VisitShortCircuitingOperator(binOp, sense: sense, stopSense: sense, stopValue: true, captureIdForResult, fallToTrueOpt, fallToFalseOpt);
+            }
+            else
+            {
+                // generate (a && b)
+                return VisitShortCircuitingOperator(binOp, sense: sense, stopSense: !sense, stopValue: false, captureIdForResult, fallToTrueOpt, fallToFalseOpt);
+            }
+        }
+
+        private IOperation VisitNullableBinaryConditionalOperator(IBinaryOperation binOp, int? captureIdForResult)
+        {
+            SpillEvalStack();
+
+            Compilation compilation = ((Operation)binOp).SemanticModel.Compilation;
+            IOperation left = binOp.LeftOperand;
+            IOperation right = binOp.RightOperand;
+            IOperation condition;
+
+            bool isAndAlso = CalculateAndOrSense(binOp, true);
+
+            // case BinaryOperatorKind.ConditionalOr:
+            //        Dim result As Boolean?
+            //
+            //        If left.GetValueOrDefault Then
```

Comment:
Might be good to add a note that, by the VB spec, if the first operand is Nothing, we evaluate the second operand anyway, but the result is guaranteed to be Nothing.

Reply:
Just so the next person isn't as confused as I was on initial read.

---
In reply to: [189756028](https://github.com/dotnet/roslyn/pull/26994#discussion_r189756028) [](ancestors = 189756028)

Reply:
> Might be good to add a note that, by the VB spec, if the first operand is Nothing, we evaluate the second operand anyway, but the result is guaranteed to be Nothing.

This is what three-value logic mean

---
In reply to: [189756507](https://github.com/dotnet/roslyn/pull/26994#discussion_r189756507) [](ancestors = 189756507,189756028)

Reply:
That confused me as well (+1 on adding a comment please).
I looked for the lowering logic for lifted Or on nullable bools (`RewriteLiftedBooleanBinaryOperator` I think) and there I saw some checks for `left.HasValue` and `right.HasValue` in the pattern. I am looking in the wrong place?

Reply:
> I am looking in the wrong place?

Probably the right place. It is not a goal to duplicate the lowering as compiler does, the goal is to accurately reflect the semantics, the order of evaluation and conditional nature of evaluation. 

---
In reply to: [190112981](https://github.com/dotnet/roslyn/pull/26994#discussion_r190112981) [](ancestors = 190112981)